dozer-ingestion/src/connectors/dozer/connector.rs

```diff
@@ -0,0 +1,235 @@
+        Self { config }
+    }
+    async fn get_client(
+        config: &NestedDozerConfig,
+    ) -> Result<InternalPipelineServiceClient<Channel>, ConnectorError> {
+        let app_server_addr =
+            Self::get_server_addr(config.grpc.as_ref().expect("grpc is required"));
+        let client = InternalPipelineServiceClient::connect(app_server_addr)
+            .await
+            .map_err(|e| {
+                ConnectorError::NestedDozerConnectorError(
+                    NestedDozerConnectorError::ConnectionError(e),
+                )
+            })?;
+        Ok(client)
+    }
+
+    async fn describe_application(&self) -> Result<DescribeApplicationResponse, ConnectorError> {
+        let mut client = Self::get_client(&self.config).await?;
+
+        let response = client
+            .describe_application(DescribeApplicationRequest {})
+            .await
+            .map_err(|e| {
+                ConnectorError::NestedDozerConnectorError(
+                    NestedDozerConnectorError::DescribeEndpointsError(e),
+                )
+            })?;
+
+        Ok(response.into_inner())
+    }
+
+    fn get_server_addr(config: &AppGrpcOptions) -> String {
+        format!("http://{}:{}", config.host, config.port)
+    }
+
+    fn get_log_options(endpoint: String, value: NestedDozerLogOptions) -> LogReaderOptions {
+        LogReaderOptions {
+            endpoint,
+            batch_size: value.batch_size,
+            timeout_in_millis: value.timeout_in_millis,
+            buffer_size: value.buffer_size,
+        }
+    }
+
+    async fn get_reader_builder(
+        &self,
+        endpoint: String,
+    ) -> Result<LogReaderBuilder, ConnectorError> {
+        let app_server_addr =
+            Self::get_server_addr(self.config.grpc.as_ref().expect("grpc is required"));
+
+        let log_options = match self.config.log_options.as_ref() {
+            Some(opts) => opts.clone(),
+            None => default_log_options().unwrap(),
+        };
+        let log_options = Self::get_log_options(endpoint, log_options);
+        let log_reader_builder = LogReaderBuilder::new(app_server_addr, log_options)
+            .await
+            .map_err(|e| {
+                NestedDozerConnectorError::ReaderError(ReaderError::ReaderBuilderError(e))
+            })?;
+        Ok(log_reader_builder)
+    }
+}
+
+async fn read_table(
+    table_idx: usize,
+    reader_builder: LogReaderBuilder,
+    ingestor: Ingestor,
+) -> Result<(), ConnectorError> {
+    let mut reader = reader_builder.build(0, None);
+    loop {
+        let (op, seq_no) = reader.next_op().await.map_err(|e| {
+            ConnectorError::NestedDozerConnectorError(NestedDozerConnectorError::ReaderError(e))
+        })?;
+        let msg = match op {
+            LogOperation::Op { op } => IngestionMessage::new_op(0, seq_no, table_idx, op),
+            LogOperation::Commit { .. } | LogOperation::SnapshottingDone { .. } => continue,
```

Comment:
We need to think about what "snapshotting" means for dozer connector. Not sending `SnapshottingDone` is not an option because downstream is relying on this message to enable certain optimizations.

The solution in my mind is like following:

`LogOperation::SnapshottingDone` should not include `connection_name`, and is only emitted once all connection affecting this log is done snapshotting.

We can leave it for another PR.